### PR TITLE
Kernel/Net: IPv6 support, other net work

### PR DESCRIPTION
**Pull request by logkos**
This PR introduces IPv6 networking support and other userland changes to work with IPv6 

In progress

### CODE REVIEW

File: Kernel/Net/Ethernet/Ethernet.h [hunk 1]
```diff
@@ -1,15 +1,17 @@
 /*
- * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2020, the SerenityOS developers.
```
Comment (Hendiadyoin1) [inline Kernel/Net/Ethernet/Ethernet.h:2]: Nope

File: Kernel/Net/Ethernet/Ethernet.h [hunk 2]
```diff
@@ -1,15 +1,17 @@
...
  *
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
 #pragma once
+#pragma GCC diagnostic ignored "-Warray-bounds"
```
Comment (Hendiadyoin1) [inline Kernel/Net/Ethernet/Ethernet.h:8]: No reason to move that up, it only applies to this file and not to the includes of it

File: Kernel/Net/Ethernet/Ethernet.h [hunk 3]
```diff
@@ -36,3 +38,10 @@ class [[gnu::packed]] EthernetFrameHeader {
 };
 
 static_assert(sizeof(EthernetFrameHeader) == 14);
+
+enum EtherType : u16 {
+    ARP = 0x0806,
+    IPv4 = 0x0800,
+    IPv6 = 0x86DD,
+};
+}
```
Comment (Hendiadyoin1) [inline Kernel/Net/Ethernet/Ethernet.h:47]: A newline here would be nice

File: Kernel/Net/IPv4/Socket.cpp [hunk 4]
```diff
@@ -234,7 +234,7 @@ ErrorOr<size_t> IPv4Socket::sendto(OpenFileDescription&, UserOrKernelBuffer cons
         if (!packet)
             return set_so_error(ENOMEM);
         routing_decision.adapter->fill_in_ipv4_header(*packet, local_address(), routing_decision.next_hop,
-            m_peer_address, (IPv4Protocol)protocol(), data_length, m_type_of_service, m_ttl);
+            m_peer_address, (Protocol)protocol(), data_length, m_type_of_service, m_ttl);
```
Comment (Hendiadyoin1) [inline Kernel/Net/IPv4/Socket.cpp:237]: while youre at it make it use c++ style casts

File: Kernel/Net/NetworkTask.cpp [hunk 5]
```diff
@@ -208,29 +207,29 @@ void handle_ipv4(EthernetFrameHeader const& eth, size_t frame_size, UnixDateTime
             update_arp_table(packet.source(), eth.source(), UpdateTable::Set);
     });
 
-    switch ((IPv4Protocol)packet.protocol()) {
-    case IPv4Protocol::ICMP:
-        return handle_icmp(eth, packet, packet_timestamp);
-    case IPv4Protocol::UDP:
+    switch ((Protocol)packet.protocol()) {
```
Comment (Hendiadyoin1) [inline Kernel/Net/NetworkTask.cpp:210]: same here

File: Kernel/Net/NetworkTask.h [hunk 6]
```diff
@@ -6,7 +6,16 @@
 
 #pragma once
 
+#include <AK/Types.h>
 namespace Kernel {
+    enum class Protocol : u16 {
```
Comment (Hendiadyoin1) [inline Kernel/Net/NetworkTask.h:11]: newline, namespace, newline

File: Kernel/Net/NetworkTask.h [hunk 7]
```diff
@@ -6,7 +6,16 @@
...
+    ICMPv4 = 1,
+    IPv4 = 4,
+    TCP = 6,
+    UDP = 17,
+    IPv6 = 41,
+    ICMPv6 = 58,
+};
 class NetworkTask {
```
Comment (Hendiadyoin1) [inline Kernel/Net/NetworkTask.h:19]: newline here

File: Kernel/Net/Ethernet/Ethernet.h (hunk 1 above)
Comment (logkos) [reply Kernel/Net/Ethernet/Ethernet.h:2]: git did that (no offense)

**Comment by AtkinsSJ** [comment]
What's the benefit of these changes? From CONTRIBUTING.md:

> Don't:
> - Use weasel-words like "refactor" or "fix" to avoid explaining what's being changed.
> - Engage in excessive "feng shui programming" by moving code around without quantifiable benefit.

I'm sure you have a reason for making these, but if you don't explain that reasoning in the commit messages or PR description then it just looks arbitrary.

**Comment by logkos** [comment]
At first glance, I thought this prep may benefit even if nothing comes out of IPv6 attempt

I've changed commit names and PR description accordingly

File: Kernel/Net/NetworkTask.h [hunk 8]
```diff
@@ -7,6 +7,7 @@
 #pragma once
 
 #include <AK/Types.h>
+
```
[the remark this answers is not in the document]
Comment (logkos) [reply Kernel/Net/NetworkTask.h:10]: probably

**Comment by logkos** [comment]
I hope this is clear now

**Pull request closed without merging**